core/src/sumeragi/network_topology.rs

```diff
@@ -125,29 +128,125 @@ impl Topology {
     pub fn leader(&self) -> &PeerId {
         &self.sorted_peers[0]
     }
+
     /// Get proxy tail's peer id.
     pub fn proxy_tail(&self) -> &PeerId {
-        &self.sorted_peers[self.min_votes_for_commit()]
+        &self.sorted_peers[self.min_votes_for_commit() - 1]
     }
+
     /// Add or remove peers from the topology.
     pub fn update_peer_list(&mut self, mut new_peers: HashSet<PeerId>) {
         self.sorted_peers.retain(|peer| new_peers.remove(peer));
         self.sorted_peers.extend(new_peers);
     }
     /// Rotate peers after each failed attempt to create a block.
     pub fn rotate_all(&mut self) {
-        self.sorted_peers.rotate_left(1);
+        self.rotate_all_n(1);
     }
+
+    /// Rotate peers n times where n is a number of failed attempt to create a block.
+    pub fn rotate_all_n(&mut self, n: usize) {
+        let len = self.sorted_peers.len();
+        if let Some(rem) = n.checked_rem(len) {
+            self.sorted_peers.rotate_left(rem);
+        }
+    }
```

Comment:
There should be a way to do this in $O(1)$

Reply:
So how this can look like imo:

1. Use circular linked list and have a pointer to current head of the list.
2. In such case `rotate_set_a` operation will be to just moving leader element to the proxy tail position which is `O(1)`.
3. `rotate_all_n` will be just moving head pointer `n` elements forward

To optimize allocation fragmentation we can use [typed_arena](https://docs.rs/typed-arena/latest/typed_arena) for node storage.

However `update_peer_list` and `lift_up_peers` might be not so optimal in such case.

Reply:
Hmm, probably situation can be even simpler than that. 
If we use observation that we construct topology once per round and then only use rotate all.
We can just store index of current leader, then `rotate_all` will be just increment of this index.

Reply:
Created #3519 for further discussion.